### PR TITLE
Modify graph and 0-D Tensor

### PR DESCRIPTION
**Pull request by lixiang007666**
此PR完成了：

- [x] 解决不支持0-dim tensor op的问题。
- [x] 添加部分0-dim测试。
- [x] 打开部分graph。
- [x] 添加ZeroDimCompatiableShape函数，将0dim tensor矫正为1dim tensor。
EX：const Shape& in_shape = ZeroDimCompatiableShape(in_tensor_desc.shape());

issue记录：https://github.com/Oneflow-Inc/OneTeam/issues/885

### CODE REVIEW

File: python/oneflow/test_utils/automated_test_util/generators.py [hunk 1]
```diff
@@ -248,9 +248,7 @@ def _generate(self, annotation):
             val = float(rng.random() * (high - low) + low)
         elif annotation == bool:
             val = random_util.choice([True, False])
-        elif annotation is None:
-            val = None
-        elif annotation is NoneType:
+        elif annotation == NoneType:
```
Comment (BBuf) [inline python/oneflow/test_utils/automated_test_util/generators.py:251]: 这里不需要改。

Comment (lixiang007666) [reply python/oneflow/test_utils/automated_test_util/generators.py]: 知道了～改好了

File: oneflow/user/ops/flatten_op.cpp [hunk 2]
```diff
@@ -26,11 +26,14 @@ namespace oneflow {
   const Shape& in_shape = in_tensor_desc.shape();
 
   CHECK_GE_OR_RETURN(start_dim, 0);
-  CHECK_LT_OR_RETURN(start_dim, in_shape.NumAxes());
+  // NOTE(lixiang): CHECK_LE_OR_RETURN for 0-dim tensor
+  CHECK_LE_OR_RETURN(start_dim, in_shape.NumAxes());
```
Comment (strint) [inline oneflow/user/ops/flatten_op.cpp:30]: 直接改放松检查降低了合理性。不过原因是0-D tensor不是合法的Tensor。

这里是不是可以做两个事情：
- 判定0-D通用工具函数
- ShapeWrapper，ShapeWrapper(Shape)可以把0-D可以把0-D的shape转成1-D，即(1, )
这样先判别0-D，然后生成ShapeWrapper，之后后面的检查都按ShapeWrapper去做检查，是不是就不用改判别条件了？

**Review by strint — COMMENTED** [review]
其它测试case的补全很完整，赞。0-D 的 check这块有优化空间，看看是否可以优化

File: oneflow/core/common/shape.cpp [hunk 3]
```diff
@@ -38,6 +38,16 @@ Shape CreateLeftExtendedShape(const ShapeView& shape, int ndims_left_extend_to)
   return Shape(std::move(dim_vec));
 }
 
+Shape ZeroDimCompatiableShape(const Shape& shape) {
+  DimVector dim_vec;
+  dim_vec = shape.dim_vec();
+  if (shape.NumAxes() == 0 && shape.elem_cnt() == 1) {
```
Comment (strint) [inline oneflow/core/common/shape.cpp:44]: 42 & 43写到这个if里面

File: oneflow/user/ops/unfold_tensor_op.cpp [hunk 4]
```diff
@@ -39,7 +39,8 @@ namespace oneflow {
   const Shape& in_shape = ctx->InputShape("x", 0);
```
Comment (strint) [inline oneflow/user/ops/unfold_tensor_op.cpp:39]: 这里矫正下，下面的check是不是不用改了

File: oneflow/core/common/shape_view.cpp [hunk 5]
```diff
@@ -27,7 +27,8 @@ ShapeView::ShapeView(const Shape& shape)
 template<typename DimT>
 int64_t ShapeViewBase<DimT>::At(int64_t index) const {
   CHECK_GE(index, 0);
-  CHECK_LT(index, num_axes_);
+  // NOTE(lixiang): CHECK_LE for 0-dim tensor
+  CHECK_LE(index, num_axes_);
```
Comment (strint) [inline oneflow/core/common/shape_view.cpp:31]: 这里是基础类。可以检查严格点：
- 非0-Dim的，走这个检查；
- 0-Dim的，可以不走这个检查

File: oneflow/user/ops/slice_op.cpp [hunk 6]
```diff
@@ -71,7 +71,8 @@ bool IsFullSlice(int64_t start, int64_t stop, int64_t step, int64_t size) {
     start = RegulateSliceStart(start, dim_size);
     stop = RegulateSliceStop(stop, dim_size);
     if (step > 0) {
-      CHECK_LT_OR_RETURN(start, stop) << "slice start must be less than stop when step > 0"
+      // NOTE(lixiang): CHECK_LE_OR_RETURN for 0 dim tensor
+      CHECK_LE_OR_RETURN(start, stop) << "slice start must be less than stop when step > 0"
```
Comment (strint) [inline oneflow/user/ops/slice_op.cpp:75]: 上面的51行的x_shape矫正下，是不是也可以不用改这个check了？

File: oneflow/user/ops/unfold_tensor_op.cpp (hunk 4 above)
Comment (lixiang007666) [reply oneflow/user/ops/unfold_tensor_op.cpp:39]: 没矫正的原因是：矫正之后会有其他的check报错。

Comment (lixiang007666) [reply oneflow/user/ops/unfold_tensor_op.cpp:39]: 这个可以后面慢慢定位

File: oneflow/core/common/shape.cpp [hunk 7]
```diff
@@ -39,9 +39,9 @@ Shape CreateLeftExtendedShape(const ShapeView& shape, int ndims_left_extend_to)
 }
 
 Shape ZeroDimCompatiableShape(const Shape& shape) {
-  DimVector dim_vec;
-  dim_vec = shape.dim_vec();
   if (shape.NumAxes() == 0 && shape.elem_cnt() == 1) {
+    DimVector dim_vec;
+    dim_vec = shape.dim_vec();
     dim_vec.emplace_back(1);
     return Shape(dim_vec);
```
Comment (strint) [inline oneflow/core/common/shape.cpp:46]: 可以简化成：
```
    DimVector dim_vec;
    dim_vec.emplace_back(1);
```

File: oneflow/core/common/shape_view.cpp [hunk 8]
```diff
@@ -27,7 +27,7 @@ ShapeView::ShapeView(const Shape& shape)
 template<typename DimT>
 int64_t ShapeViewBase<DimT>::At(int64_t index) const {
   CHECK_GE(index, 0);
-  CHECK_LT(index, num_axes_);
+  if (!(this->NumAxes() == 0 && this->elem_cnt() == 1)) { CHECK_LT(index, num_axes_); }
```
Comment (strint) [inline oneflow/core/common/shape_view.cpp:30]: else {
   CHECK(index == 0);
}

**Review by strint — APPROVED** [review]

**Review by BBuf — APPROVED** [review]

File: python/oneflow/test_utils/automated_test_util/generators.py [hunk 9]
```diff
@@ -281,7 +281,7 @@ def __init__(
         dtype=float,
     ):
         if ndim is None:
-            ndim = random(1, 6)
+            ndim = random(0, 6)
```
Comment (strint) [inline python/oneflow/test_utils/automated_test_util/generators.py:284]: 是不是这里，会生成 scalar tensor，然后有case 过不了 sclar

Comment (wyg1997) [inline python/oneflow/test_utils/automated_test_util/generators.py:284]: 这里会产生 size 完全为空的 Tensor，导致了 flow.min 出错，复现代码：


```python
import oneflow as flow

a = flow.ones([]).requires_grad_()
b = flow.min(a)
c = b.sum().backward()
```

这里我需要优化一下 autograd 那边的报错信息

Comment (wyg1997) [reply python/oneflow/test_utils/automated_test_util/generators.py:284]: 这里先改回去，最小的 dim 为 1 吧。

**Comment by wyg1997** [comment]
现在 Scalar leaf Tensor调用 backward 的问题解决了，上面那个讨论（ https://github.com/Oneflow-Inc/oneflow/pull/7208#discussion_r784771597 ）就也没问题了，是否可以把 ndim 那个改成 0，产生 Scalar Tensor 做测试了？ @strint

File: oneflow/core/common/shape_view.cpp [hunk 10]
```diff
@@ -27,7 +27,11 @@ ShapeView::ShapeView(const Shape& shape)
 template<typename DimT>
 int64_t ShapeViewBase<DimT>::At(int64_t index) const {
   CHECK_GE(index, 0);
-  CHECK_LT(index, num_axes_);
+  if (!(this->NumAxes() == 0 && this->elem_cnt() == 1)) {
+    CHECK_LT(index, num_axes_);
+  } else {
+    CHECK(index == 0);
+  }
   return ptr_[index];
```
Comment (daquexian) [inline oneflow/core/common/shape_view.cpp:35]: 我在做 ShapeView 的重构，现在 0-d tensor 的 shape 是虽然 NumAxes() == 0，但其实有一个元素吗？

Comment (lixiang007666) [reply oneflow/core/common/shape_view.cpp:35]: 我理解相当于scalar tensor，所以有一个元素。

Comment (wyg1997) [reply oneflow/core/common/shape_view.cpp:35]: 是这样的，0dim Tensor 的 NumAxes 为 0，但 elem_cnt 是 1。0size tensor 的 NumAxes 不一定是 0，但 elem_cnt 为 0。

Comment (daquexian) [reply oneflow/core/common/shape_view.cpp:35]: > 我理解相当于scalar tensor，所以有一个元素。

嗯嗯我是说 Shape 不是 Tensor 本身。因为我圈的这段代码的逻辑是 “如果 shape 的 NumAxes() == 0，就检查 index 是否为 0，并返回 ptr_[0]”，但 “NumAxes() == 0” 和 “ptr_[0] 存在” 是有点矛盾的

Comment (lixiang007666) [reply oneflow/core/common/shape_view.cpp:35]: 哦哦哦～我猜是不是取第0维的大小（ptr_[0]）就是0？

Comment (lixiang007666) [reply oneflow/core/common/shape_view.cpp:35]: 还是说这种情况下，不能return ptr_[0]。

Comment (daquexian) [reply oneflow/core/common/shape_view.cpp:35]: 按 0-d tensor 的语义，shape 应该是一个空的 vector，shape.At(0) 就不是很合理，应该会有越界访问才对😂 如果之前没有出过 bug，可能是其它地方也有类似的改动叠加起来得到了错错得对的效果，所以就想了解下 graph 里为 0-d tensor 构造 header 的处理是怎样的

Comment (lixiang007666) [reply oneflow/core/common/shape_view.cpp:35]: 貌似有些涉及到 0d tensor 的输入 shape 的地方都改成了 1d，这样就可以用shape.At(0)了。不过这段代码确实不合理。👀

Comment (lixiang007666) [reply oneflow/core/common/shape_view.cpp:35]: 这个 PR 里加了 ZeroDimCompatiableShape 这个函数。

Comment (daquexian) [reply oneflow/core/common/shape_view.cpp:35]: 提交了代码 https://github.com/Oneflow-Inc/oneflow/pull/8422/commits/ba26bbe17d1aa99b895d75bb715c023589ea1681 ，可以 review 下 @lixiang007666 @strint @BBuf
